src/main/java/com/gargoylesoftware/htmlunit/WebResponseFromCache.java

```diff
@@ -25,6 +28,19 @@
 class WebResponseFromCache extends WebResponseWrapper {
 
     private final WebRequest request_;
+    private final List<NameValuePair> responseHeaders_;
+
+    /**
+     * Wraps the provide response for the given request
+     * @param cachedResponse the response from cache
+     * @param overwriteHeaders the list of headers to overwrite cachedResponse headers
+     * @param currentRequest the new request
+     */
+    WebResponseFromCache(final WebResponse cachedResponse, final List<NameValuePair> overwriteHeaders, final WebRequest currentRequest) {
+        super(cachedResponse);
+        request_ = currentRequest;
+        responseHeaders_ = overwriteHeaders;
```

Comment:
needs to be immutable
```suggestion
        responseHeaders_ = Collections.unmodifiableList(overwriteHeaders);
```